File: app/src/main/java/uk/co/glass_software/android/cache_interceptor/demo/ExpandableListAdapter.java
package uk.co.glass_software.android.cache_interceptor.demo;


import android.content.Context;
import android.graphics.Typeface;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;

import io.reactivex.Observable;
import uk.co.glass_software.android.cache_interceptor.demo.model.JokeResponse;
import uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheToken;
import uk.co.glass_software.android.cache_interceptor.interceptors.error.ApiError;
import uk.co.glass_software.android.cache_interceptor.utils.Action;
import uk.co.glass_software.android.cache_interceptor.utils.Callback;

class ExpandableListAdapter extends BaseExpandableListAdapter {
    
    private final Callback<String> jokeCallback;
    private final Action onComplete;
    private final LinkedList<String> headers;
    private final LinkedList<String> logs;
    private final LinkedHashMap<String, List<String>> children;
    private final LayoutInflater inflater;
    private final SimpleDateFormat simpleDateFormat;
    
    ExpandableListAdapter(Context context,
                          Callback<String> jokeCallback,
                          Action onComplete) {
        this.jokeCallback = jokeCallback;
        this.onComplete = onComplete;
        headers = new LinkedList<>();
        logs = new LinkedList<>();
        children = new LinkedHashMap<>();
        inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        simpleDateFormat = new SimpleDateFormat("hh:mm:ss");
    }
    
    void loadJoke(Observable<? extends JokeResponse> observable) {
        headers.clear();
        children.clear();
        logs.clear();
        long start = System.currentTimeMillis();
        observable.doOnComplete(() -> onComplete(start))
                  .subscribe(joke -> onJokeReady(start, joke));
    }
    
    private void onJokeReady(long start,
                             JokeResponse jokeResponse) {
        CacheToken<JokeResponse> cacheToken = jokeResponse.getMetadata().getCacheToken();
        String ellapsed = cacheToken.getStatus() + ", " + (System.currentTimeMillis() - start) + "ms";
        List<String> info = new ArrayList<>();
        String header;
        
        if (jokeResponse.getMetadata().hasError()) {
            header = "An error occurred: " + ellapsed;
            
            ApiError error = jokeResponse.getMetadata().getError();
            info.add("Description: " + error.getDescription());
            info.add("Message: " + error.getMessage());
            info.add("Cause: " + error.getCause());
        }
        else {
            String joke = String.valueOf(Html.fromHtml(jokeResponse.getValue().getJoke()));
            jokeCallback.call(joke);
            header = ellapsed;
            
            info.add("Cache token status: " + cacheToken.getStatus());
            info.add("Cache token cache date: " + simpleDateFormat.format(cacheToken.getCacheDate()));
            info.add("Cache token expiry date: "
                     + simpleDateFormat.format(cacheToken.getExpiryDate())
                     + " (TTL: "
                     + (int) (jokeResponse.getTtlInMinutes() * 60f)
                     + "s)"
            );
            
            info.add("Joke: " + joke);
        }
        
        headers.add(header);
        children.put(header, info);
        
        notifyDataSetChanged();
    }
    
    void log(String output) {
        logs.addLast(output);
    }
    
    private void onComplete(long start) {
        String header = "Log output (total: " + (System.currentTimeMillis() - start) + "ms)";
        headers.add(header);
        children.put(header, logs);
        notifyDataSetChanged();
        onComplete.act();
    }
    
    @Override
    public Object getChild(int groupPosition,
                           int childPosition) {
        String key = headers.get(groupPosition);
        return children.get(key).get(childPosition);
    }
    
    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }
    
    @Override
    public View getGroupView(int groupPosition,
                             boolean isExpanded,
                             View convertView,
                             ViewGroup parent) {
        String headerTitle = (String) getGroup(groupPosition);
        if (convertView == null) {
            convertView = inflater.inflate(R.layout.list_group, null);
        }
        
        TextView lblListHeader = convertView.findViewById(R.id.lblListHeader);
        lblListHeader.setTypeface(null, Typeface.BOLD);
        lblListHeader.setText(headerTitle);
        
        return convertView;
    }
    
    @Override
    public View getChildView(int groupPosition,
                             final int childPosition,
                             boolean isLastChild,
                             View convertView,
                             ViewGroup parent) {
        final String childText = (String) getChild(groupPosition, childPosition);
        
        if (convertView == null) {
            convertView = inflater.inflate(R.layout.list_item, null);
        }
        
        TextView txtListChild = convertView.findViewById(R.id.lblListItem);
        
        txtListChild.setText(childText);
        return convertView;
    }
    
    @Override
    public int getChildrenCount(int groupPosition) {
        return children.get(headers.get(groupPosition)).size();
    }
    
    @Override
    public Object getGroup(int groupPosition) {
        return headers.get(groupPosition);
    }
    
    @Override
    public int getGroupCount() {
        return headers.size();
    }
    
    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }
    
    @Override
    public boolean hasStableIds() {
        return false;
    }
    
    @Override
    public boolean isChildSelectable(int groupPosition,
                                     int childPosition) {
        return false;
    }
}

File: app/src/main/java/uk/co/glass_software/android/cache_interceptor/demo/ExpandableListAdapter.kt
package uk.co.glass_software.android.cache_interceptor.demo


import android.content.Context
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.TextView
import uk.co.glass_software.android.cache_interceptor.configuration.CacheInstruction
import uk.co.glass_software.android.cache_interceptor.configuration.CacheInstruction.Operation.Expiring
import uk.co.glass_software.android.cache_interceptor.demo.model.CatFactResponse
import uk.co.glass_software.android.cache_interceptor.interceptors.internal.cache.token.CacheStatus
import uk.co.glass_software.android.cache_interceptor.interceptors.internal.cache.token.CacheStatus.*
import java.text.SimpleDateFormat
import java.util.*

internal class ExpandableListAdapter(context: Context)
    : BaseExpandableListAdapter() {

    private val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    private val simpleDateFormat = SimpleDateFormat("MM/dd/YY hh:mm:ss")

    private val headers: LinkedList<String> = LinkedList()
    private val logs: LinkedList<String> = LinkedList()
    private val children: LinkedHashMap<String, List<*>> = LinkedHashMap()

    private var callStart = 0L

    fun onStart(instruction: CacheInstruction) {
        headers.clear()
        children.clear()
        logs.clear()

        callStart = System.currentTimeMillis()

        val header = "Retrofit Call"
        headers.add(header)
        children[header] = listOf(instruction)

        notifyDataSetChanged()
    }

    fun showCatFact(catFactResponse: CatFactResponse) {
        val metadata = catFactResponse.metadata
        val cacheToken = metadata.cacheToken
        val exception = metadata.exception
        val operation = cacheToken.instruction.operation.type
        val callDuration = metadata.callDuration

        val elapsed = "${operation.name} -> ${cacheToken.status} (${callDuration.total}ms)"
        val duration = "Call duration: disk = ${callDuration.disk}ms, network = ${callDuration.network}ms, total = ${callDuration.total}ms"

        val info = ArrayList<String>()
        val header: String

        info.add("Cache token instruction: $operation")
        info.add("Cache token status: ${cacheToken.status} (coming from ${getOrigin(cacheToken.status)})")

        if (exception != null) {
            header = "An error occurred: $elapsed"

            info.add("Description: " + exception.description)
            info.add("Message: " + exception.message)
            info.add("Cause: " + exception.cause)
            info.add(duration)
        } else {
            header = elapsed

            info.add("Cache token cache date: " + simpleDateFormat.format(cacheToken.cacheDate))
            info.add(duration)

            if (operation is Expiring) {
                info.add("Cache token expiry date: "
                        + simpleDateFormat.format(cacheToken.expiryDate)
                        + " (TTL: "
                        + (operation.durationInMillis?.times(1000)?.toInt() ?: "N/A")
                        + "s)"
                )
            }
        }

        headers.add(header)
        children[header] = info

        val catFactHeader = "Cat Fact (${catFactResponse.metadata.cacheToken.status})"
        headers.add(catFactHeader)
        children[catFactHeader] = listOf(catFactResponse.fact ?: "N/A")

        notifyDataSetChanged()
    }

    fun onComplete() {
        val header = "Log Output (Total: " + (System.currentTimeMillis() - callStart) + "ms)"
        headers.add(header)
        children[header] = logs
        notifyDataSetChanged()
    }

    private fun getOrigin(status: CacheStatus) =
            when (status) {
                INSTRUCTION -> "instruction"
                NOT_CACHED,
                FRESH,
                REFRESHED -> "network"
                CACHED,
                STALE,
                COULD_NOT_REFRESH -> "disk"
                EMPTY -> "network or disk"
            } + ", " + (if (status.isFinal) "final" else "non-final")

    fun log(output: String) {
        logs.addLast(output)
    }

    override fun getChild(groupPosition: Int,
                          childPosition: Int) = headers[groupPosition].let {
        children[it]!![childPosition]
    }

    override fun getChildId(groupPosition: Int, childPosition: Int) = childPosition.toLong()

    override fun getGroupView(groupPosition: Int,
                              isExpanded: Boolean,
                              convertView: View?,
                              parent: ViewGroup) =
            (convertView ?: inflater.inflate(R.layout.list_group, parent, false))
                    .apply {
                        findViewById<TextView>(R.id.listHeader).apply {
                            setTypeface(null, Typeface.BOLD)
                            text = getGroup(groupPosition)
                        }
                    }!!

    override fun getChildView(groupPosition: Int,
                              childPosition: Int,
                              isLastChild: Boolean,
                              convertView: View?,
                              parent: ViewGroup): View =
            (convertView ?: inflater.inflate(R.layout.list_item, parent, false))
                    .apply {
                        val child = getChild(groupPosition, childPosition)
                        val text = findViewById<TextView>(R.id.listItem)
                        val instruction = findViewById<InstructionView>(R.id.instruction)

                        if (child is String) {
                            text.visibility = View.VISIBLE
                            instruction.visibility = View.GONE
                            text.text = child
                        } else if (child is CacheInstruction) {
                            text.visibility = View.GONE
                            instruction.visibility = View.VISIBLE
                            instruction.setInstruction(child)
                        }
                    }

    override fun getChildrenCount(groupPosition: Int) = children[headers[groupPosition]]!!.size

    override fun getGroup(groupPosition: Int) = headers[groupPosition]

    override fun getGroupCount() = headers.size

    override fun getGroupId(groupPosition: Int) = groupPosition.toLong()

    override fun hasStableIds() = false

    override fun isChildSelectable(groupPosition: Int,
                                   childPosition: Int) = false
}
